### PR TITLE
Fix display units of moments in velocity

### PR DESCRIPTION
**Pull request by rosteen**
This now converts the spectral axis of the input cube to velocity prior to sending it through the moment calculation rather than attempting to convert the result afterward, thus fixing the display units of the result. 

I believe that the reported unrealistic negative values for moment 2 are an artifact of doing continuum subtraction on pixels that are essentially noise around the continuum - the calculation is `np.sum(flux * (dispersion - m1) ** order, axis=axis) / m0`, so even though the dispersion term is getting squared, with negative flux values (especially far from m1) you can end up with a negative second moment in those cases. Everything looks ok when you do the moment calculation without continuum subtraction. Here's an example of the spectrum at a pixel that results in a negative second moment at that pixel:

Screenshot 2024-01-18 at 10 54 50 AM

### CODE REVIEW

File: jdaviz/configs/cubeviz/plugins/moment_maps/moment_maps.py [hunk 1]
```diff
@@ -243,20 +242,25 @@ def calculate_moment(self, add_data=True):
...
-            in_velocity = c*relative_wavelength
-            if self.output_unit_selected.lower() == "velocity^n":
-                in_velocity = np.power(in_velocity, self.n_moment)
-            self.moment = CCDData(in_velocity, wcs=data_wcs)
+            slab_sa = slab.spectral_axis.to("km/s", doppler_convention="relativistic",
+                                            doppler_rest=ref_wavelength)
+            slab = Spectrum1D(slab.flux, slab_sa)
+
+        # Finally actually calculate the moment
+        self.moment = analysis.moment(slab, order=n_moment).T
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/moment_maps/moment_maps.py:258]: I feel like intermediate calculations shouldn't change the actual instance attribute, but it is just a personal preference, so no biggie.

File: jdaviz/configs/cubeviz/plugins/moment_maps/tests/test_moment_maps.py [hunk 2]
```diff
@@ -165,7 +165,7 @@ def test_moment_velocity_calculation(cubeviz_helper, spectrum1d_cube, tmpdir):
     label_mouseover = cubeviz_helper.app.session.application._tools['g-coords-info']
     label_mouseover._viewer_mouse_event(uncert_viewer, {'event': 'mousemove',
                                                         'domain': {'x': 0, 'y': 0}})
-    assert label_mouseover.as_text() == ("Pixel x=00.0 y=00.0 Value -4.14382e+05 m / s",
+    assert label_mouseover.as_text() == ("Pixel x=00.0 y=00.0 Value -4.14668e+02 km / s",
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/moment_maps/tests/test_moment_maps.py:168]: Why did the decimal value also change?

File: jdaviz/configs/cubeviz/plugins/moment_maps/tests/test_moment_maps.py [hunk 3]
```diff
@@ -174,10 +174,11 @@ def test_moment_velocity_calculation(cubeviz_helper, spectrum1d_cube, tmpdir):
     mm.calculate_moment()
 
     label_mouseover._viewer_mouse_event(uncert_viewer, {'event': 'mousemove',
-                                                        'domain': {'x': 0, 'y': 0}})
-    assert label_mouseover.as_text() == ("Pixel x=00.0 y=00.0 Value -2.99792e+08 m / s",
-                                         "World 13h39m59.9731s +27d00m00.3600s (ICRS)",
-                                         "204.9998877673 27.0001000000 (deg)")
+                                                        'domain': {'x': 1, 'y': 1}})
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/moment_maps/tests/test_moment_maps.py:177]: Why change it to `x=1 y=1`? Hard to compare the diff to see if results are the same or not before/after patching.

Comment (rosteen) [reply jdaviz/configs/cubeviz/plugins/moment_maps/tests/test_moment_maps.py:177]: The new moment 2 values are indeed different, that calculation was incorrect before. And I switched to (1,1) because the new value at 0 is 0, which is boring 🙂 .

Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/moment_maps/tests/test_moment_maps.py:177]: That different, huh? I wonder how many scientists trusted the wrong values blindly. 😱 

What about the moment 1 differences above?

File: CHANGES.rst [hunk 4]
```diff
@@ -87,6 +87,8 @@ Cubeviz
 - Fixes Spectral Extraction's assumptions of one data per viewer, and flux data only in
   flux-viewer/uncertainty data only in uncert-viewer. [#2646]
 
+- Fixed incorrect output units when Velocity is selected in Moment Maps plugin. [#2665]
```
Comment (pllim) [inline CHANGES.rst:90]: Maybe this entry needs to be more descriptive: Which moment was wrong and how.

Comment (rosteen) [reply CHANGES.rst:90]: I just realized this doesn't need a separate changelog entry, it's a fix to a new 3.9 feature.

Comment (pllim) [reply CHANGES.rst:90]: Oh, phew!!!

Comment (pllim) [reply CHANGES.rst]: You might still want to attach this PR number to the original entry.

File: jdaviz/configs/cubeviz/plugins/moment_maps/tests/test_moment_maps.py (hunk 2 above)
Comment (rosteen) [reply jdaviz/configs/cubeviz/plugins/moment_maps/tests/test_moment_maps.py:168]: I _think_ that I'm not terribly surprised that there's a small (<0.1%) difference between my manual conversion of the result vs converting the spectral axis of each spaxel beforehand. Likely due to my manual calculation before using the simpler optical equation, vs now using the full relativistic convention in the conversion through specutils.

**Review by pllim — APPROVED** [review]
Code looks like code. Test updates are explained satisfactorily.

Do we need Patrick to do science verification and be the second approver?

**Review by PatrickOgle — APPROVED** [review]
The problems with units and the moment2 calculation are both fixed. 

I checked the peak wavelengths of the ‘fastest’ pixels in the map, and they do match the velocities in the moment 0 map, to within ~20 km/s. (To do this, I selected single spaxels with the spaxel tool and used the slice tool to find the peak of line in these spaxel spectra,  then computed velocity = c*(lambda_peak - lambda_ref)/lambda.)  The ~20 km/s differences are related to the line shape and the accuracy to which I can center the slice marker.

Bottom line--I see no gross problems with the moment calculation for single spaxels.  The science is ‘good’ in this case, and shows that stuff is moving faster around a black hole  that is more massive than the ground measurements indicated.  Makes sense, since JWST has better resolution.
